esp32_ulp/opcodes.py

```diff
@@ -267,6 +268,22 @@ def make_ins(layout):
 ARG = namedtuple('ARG', ('type', 'value', 'raw'))
 
 
+def eval_arg(arg):
+    parts = []
+    for token in split_tokens(arg):
+        if symbols.has_sym(token):
+            _, _, sym_addr = symbols.get_sym(token)
+            parts.append(str(sym_addr))
+        else:
+            parts.append(token)
+    parts = "".join(parts)
+    try:
+        parts = str(eval(parts))
```

Comment:
**Uhoh, guess this is "assembler injection" (see "sql injection").**

Reply:
Also, this means that the expression syntax is the same as for python.

Is that compatible to the usual assemblers.

Reply:
I did a fair bit of reading up here originally, because I tend to be one of the first to call `eval() == evil()`. In the end I decided that this might be a worthwhile trade-off, because for the expressions I could see being evaluated here the syntax is close enough, and in terms of misuse I found the risk quite low, as the code that gets assembled would be quite short, and python code would be very obvious (and believing code off the net should never be trusted without review anyway).

Do you feel that this is a reasonable assumption, or would you prefer we go a safer route? I will say that I would prefer a different solution, however the only one I could think of for now was to implement the expression parser / evaluator, which would add a lot of code and it may use more memory, thus reducing the size of things we can assemble on-device, and since MicroPython had a way to evaluate things, I used that.

Anyway, about "expression syntax": I had found this: https://pypi.org/project/pcpp/#known-bugs-ordered-from-worst-to-least-worst - a similar project to create a preprocessor. They had concluded that using Python to evaluate expressions worked surprisingly well and I considered their listed edge cases acceptable (perhaps this should be documented).

I see a few ways forward:
1. Do nothing. Accept the risk/imperfection
2. Add a description/warning/listing-of-edge-cases to the documentation
3. Add validation - i.e. tokenise the string-to-be-evaluated and check that it consists only of numbers, brackets, supported arithmetic/logical operators or known symbols.
4. Implement a proper parser/evaluator that does away entirely with calling eval().

Which would you prefer we pick or do you see other options? I am starting to like option 3, which I had not thought of before now.

Reply:
I just implemented option 3, as it felt easy enough to do. I have pushed it "on-top" for now (commit dd3a04b), but I intend to squash the change later, once I know you are happy with this option. (hmm, just noticed I don't have a test for the changed "eval_arg". I'll add it)

Reply:
sounds reasonable.